rollbar-reactive-streams-reactor/src/main/java/com/rollbar/reactivestreams/notifier/sender/http/ReactorAsyncHttpClient.java

```diff
@@ -0,0 +1,174 @@
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.nio.charset.StandardCharsets;
+import java.util.AbstractMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.http.client.HttpClientResponse;
+import reactor.netty.resources.ConnectionProvider;
+import reactor.netty.transport.ProxyProvider;
+
+/**
+ * <a href="https://projectreactor.io/">Project Reactor</a> {@link AsyncHttpClient} implementation.
+ */
+public class ReactorAsyncHttpClient implements AsyncHttpClient {
+  private final AtomicInteger runningRequests;
+  private final HttpClient httpClient;
+
+  ReactorAsyncHttpClient(Builder builder) {
+    this.runningRequests = new AtomicInteger(0);
+
+    HttpClient httpClient;
+    if (builder.connectionProvider != null) {
+      httpClient = HttpClient.create(builder.connectionProvider);
+    } else {
+      httpClient = HttpClient.create();
+    }
+
+    ProxyProvider.Proxy proxyType = getProxyType(builder.proxy);
+    InetSocketAddress proxyAddress = getProxyAddress(proxyType, builder.proxy);
+
+    if (proxyType != null) {
+      httpClient = httpClient.proxy(
+          typeSpec -> typeSpec.type(proxyType).address(proxyAddress));
+    }
+
+    this.httpClient = httpClient.compress(true);
+  }
+
+  @Override
+  public Publisher<AsyncHttpResponse> send(AsyncHttpRequest httpRequest) {
+
+    ByteBuf buffer = ByteBufAllocator.DEFAULT.buffer();
+    buffer.writeCharSequence(httpRequest.getBody(), StandardCharsets.UTF_8);
+    Mono<ByteBuf> buf = Mono.just(buffer);
+
+    return httpClient
+        .headers(entries -> {
+          for (Map.Entry<String, String> header : httpRequest.getHeaders()) {
+            entries.add(header.getKey(), header.getValue());
+          }
+        })
+        .post()
+        .uri(httpRequest.getUrl())
+        .send(buf)
+        .responseSingle((res, content) ->
+            content.map(body -> new AbstractMap.SimpleEntry<>(res, body)))
+        .doOnSubscribe(sig -> runningRequests.incrementAndGet())
+        .doAfterTerminate(runningRequests::decrementAndGet)
+        .map(ReactorAsyncHttpClient::mapResponse);
+  }
+
+  private static AsyncHttpResponse mapResponse(
+      AbstractMap.SimpleEntry<HttpClientResponse, ByteBuf> response) {
+    return new AsyncHttpResponse() {
+      @Override
+      public int getStatusCode() {
+        return response.getKey().status().code();
+      }
+
+      @Override
+      public Iterable<Map.Entry<String, String>> getHeaders() {
+        return response.getKey().responseHeaders();
+      }
+
+      @Override
+      public String getBody() {
+        return response.getValue().toString(StandardCharsets.UTF_8);
+      }
+    };
+  }
+
+  @Override
+  public void close(boolean wait) {
+    while (runningRequests.get() > 0) {
+      try {
+        Thread.sleep(50);
```

Comment:
Why `50`?

Reply:
High enough to avoid high CPU usage, low enough to return quickly after the requests are completed.

Reply:
The issue here is that `close()` is related to the `AutoCloseable` interface (where `AutoCloseable`'s `close` calls this overload) and that is a sync interface. So if we want to support that, we need to block. 

The solution is probably to not support `AutoCloseable`. This removes some functionality at the language level, such as [try with resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) but that's the compromises that need to be made in Java where async is not part of the language.

Reply:
Since the Apache HTTP client (which we use in our Java 7 implementation) implements a blocking close method, and that would have been quite a hassle to make non-blocking, I've instead kept the blocking implementation here, but instead of sleeping we now rely on Project Reactor's internal blocking implementation. 

That implementation supports a check that will cause the method to throw if it's invoked from one of the Reactor threads. This method is the result of Notifier.close -> Sender.close -> HTTPClient.close, so it should only be called during application shutdown, where there is no alternative to blocking if the user wants to wait for pending operations to complete.